src/internal/arg.c

```diff
@@ -84,3 +85,107 @@ sexp* rlang_ext2_is_missing(sexp* _call, sexp* _op, sexp* args, sexp* env) {
 
   return r_lgl(r_eval(r_x_sym, env) == r_missing_sym);
 }
+
+static sexp* stop_arg_match_call = NULL;
+static sexp* arg_nm_sym = NULL;
+
+void arg_match0_abort(const char* msg, sexp* env);
+sexp* rlang_ext2_arg_match0(sexp* _call, sexp* _op, sexp* args, sexp* env) {
+  args = r_node_cdr(args);
+
+  sexp* arg =    r_node_car(args); args = r_node_cdr(args);
+  sexp* values = r_node_car(args);
+
+  if (r_typeof(arg) != r_type_character) {
+    arg_match0_abort("`%s` must be a character vector.", env);
+  }
+  if (r_typeof(values) != r_type_character) {
+    r_abort("`values` must be a character vector.");
+  }
+
+  r_ssize len = r_length(arg);
+  r_ssize len_values = r_length(values);
+  if (len != 1 && len != len_values) {
+    arg = KEEP(r_str_as_character(r_chr_get(arg, 0)));
+    sexp* arg_nm = KEEP(r_eval(arg_nm_sym, env));
+    r_eval_with_xyz(stop_arg_match_call, r_base_env, arg, values, arg_nm);
```

Comment:
Eval in the base env? Maybe instead you could eval in rlang's package namespace? Then `stop_arg_match_call` wouldn't have to be `rlang:::` I think?

Reply:
This pattern is used across rlang, I don't find `r_rlang_env` or similar symbols in the code.

Reply:
You can use `rlang_ns_env`:

```
src/internal/attr.c:152:  return r_eval_with_xy(as_function_call, rlang_ns_env, x, env);
src/internal/dots.c:12:  return r_eval_with_x(as_label_call, rlang_ns_env, x);
src/internal/dots.c:208:      return r_eval_with_x(rlang_as_list_call, rlang_ns_env, x);
src/internal/dots.c:214:      return r_eval_with_x(rlang_as_list_call, rlang_ns_env, x);
src/internal/dots.c:219:    return r_eval_with_x(rlang_as_list_call, rlang_ns_env, x);
src/internal/eval-tidy.c:346:  sexp* out = r_eval_with_x(call, rlang_ns_env, mask);
src/lib/env.c:141:  r_eval_with_xyz(poke_lazy_call, rlang_ns_env, name, env, eval_env);
src/lib/cnd.c:98:  r_eval_with_x(call, r_ns_env("rlang"), msg);
```